analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Pattern.cs

```diff
@@ -34,9 +35,30 @@ internal static class Pattern
             ? context.SetOperationConstraint(newConstraint)
             : context.State;
 
+        public static ProgramState Process(SymbolicContext context, IRecursivePatternOperationWrapper recursive) =>
+            SetDeclarationNotNull(context, recursive.DeclaredSymbol);
+
+        public static ProgramState Process(SymbolicContext context, IDeclarationPatternOperationWrapper declaration) =>
+            SetDeclarationNotNull(context, declaration.DeclaredSymbol);
+
         private static BoolConstraint PatternBoolConstraint(SymbolicValue value, BoolConstraint pattern) =>
             value.HasConstraint<BoolConstraint>()
                 ? BoolConstraint.From(value.HasConstraint(pattern))
                 : null; // We cannot take conclusive decision
+
+        private static ProgramState SetDeclarationNotNull(SymbolicContext context, ISymbol declaredSymbol)
+        {
+            if (declaredSymbol == null)
+            {
+                return context.State;
+            }
+            else
+            {
+                var state = context.Operation.Parent.AsIsPattern() is { } parentIsPattern && parentIsPattern.Value.TrackedSymbol() is { } sourceSymbol
+                    ? context.State.SetSymbolValue(declaredSymbol, context.State[sourceSymbol])  // ToDo: MMF-2563 should define relation between tested and declared symbol
+                    : context.State;
+                return state.SetSymbolConstraint(declaredSymbol, ObjectConstraint.NotNull);
```

Comment:
I think this is wrong for `var ` declaration pattern:
`if (a is var b)`

From the spec:
>  In other words, a match to a var pattern always succeeds with a simple_designation.

Reply:
Good catch